platform/macos/joypad_macos.mm

```diff
@@ -0,0 +1,523 @@
+							CHHapticPatternKeyParameterID : CHHapticEventParameterIDHapticIntensity,
+							CHHapticPatternKeyParameterValue : [NSNumber numberWithFloat:p_strong_magnitude]
+						},
+					],
+				},
+			},
+		],
+	};
+	NSError *error;
+	CHHapticPattern *pattern = [[CHHapticPattern alloc] initWithDictionary:hapticDict error:&error];
+	return pattern;
+}
+
+void JoypadMacOS::joypad_vibration_start(Joypad *p_joypad, float p_weak_magnitude, float p_strong_magnitude, float p_duration, uint64_t p_timestamp) {
+	if (!p_joypad.force_feedback || p_weak_magnitude < 0.f || p_weak_magnitude > 1.f || p_strong_magnitude < 0.f || p_strong_magnitude > 1.f) {
+		return;
+	}
+
+	if (p_joypad.pattern_player != nil) {
+		joypad_vibration_stop(p_joypad, p_timestamp);
+	}
+
+	// Gets the default vibration pattern and creates a player with it.
+	NSError *error;
+	CHHapticPattern *pattern = get_vibration_pattern(p_weak_magnitude, p_strong_magnitude, p_duration);
+	id<CHHapticPatternPlayer> player = [p_joypad.motion_engine createPlayerWithPattern:pattern error:&error];
+
+	p_joypad.pattern_player = player;
+
+	// When all players have stopped for an engine, stop the engine.
+	[p_joypad.motion_engine notifyWhenPlayersFinished:^CHHapticEngineFinishedAction(NSError *_Nullable error) {
+		return CHHapticEngineFinishedActionStopEngine;
+	}];
+
+	// Starts the engine and execute the vibration pattern
+	[p_joypad.motion_engine startWithCompletionHandler:^(NSError *returnedError) {
+		NSError *error;
+		[player startAtTime:0 error:&error];
+		p_joypad.ff_effect_timestamp = p_timestamp;
+	}];
+}
+
+void JoypadMacOS::joypad_vibration_stop(Joypad *p_joypad, uint64_t p_timestamp) {
+	if (!p_joypad.force_feedback) {
+		return;
+	}
+	if (p_joypad.pattern_player == nil) {
+		return;
+	}
+
+	NSError *error;
+	[p_joypad.pattern_player stopAtTime:0 error:&error];
+	p_joypad.pattern_player = nil;
+	p_joypad.ff_effect_timestamp = p_timestamp;
+}
+
+@interface JoypadMacOSObserver ()
+
+@property(assign, nonatomic) BOOL isObserving;
+@property(assign, nonatomic) BOOL isProcessing;
+@property(strong, nonatomic) NSMutableDictionary<NSNumber *, Joypad *> *connectedJoypads;
+@property(strong, nonatomic) NSMutableArray<Joypad *> *joypadsQueue;
+
+@end
+
+@implementation JoypadMacOSObserver
+
+- (instancetype)init {
+	self = [super init];
+
+	if (self) {
+		[self godot_commonInit];
+	}
+
+	return self;
+}
+
+- (void)godot_commonInit {
+	self.isObserving = NO;
+	self.isProcessing = NO;
+}
+
+- (void)startProcessing {
+	self.isProcessing = YES;
+
+	for (GCController *controller in self.joypadsQueue) {
+		[self addMacOSJoypad:controller];
+	}
+
+	[self.joypadsQueue removeAllObjects];
+}
+
+- (void)startObserving {
+	if (self.isObserving) {
+		return;
+	}
+
+	self.isObserving = YES;
+
+	self.connectedJoypads = [NSMutableDictionary dictionary];
+	self.joypadsQueue = [NSMutableArray array];
+
+	// get told when controllers connect, this will be called right away for
+	// already connected controllers
```

Comment:
```suggestion
	// Get told when controllers connect, this will be called right away for
	// already connected controllers.
```
Per the [comment style](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide).